cleangoldenami/main.go

```diff
@@ -96,40 +112,74 @@ func main() {
 							x86Images = append(x86Images[:imageToDeleteIndex], x86Images[imageToDeleteIndex+1:]...)
 						}
 					}
-					if !*dryRun {
-						fmt.Printf("Region %v image capacity is currently at %v out of %v:\n", regionName, len(images), quota)
-					} else {
-						fmt.Printf("Region %v image capacity is currently at %v out of %v - would delete the following AMIs:\n", regionName, len(images), quota)
+					drr := deregisterRegionResult{
+						region:                 regionName,
+						currentImages:          len(images),
+						quota:                  quota,
+						deregisterImageResults: make([]deregisterImageResult, len(imagesToDelete)),
 					}
-					for _, image := range imagesToDelete {
-						for _, t := range image.Tags {
-							if *t.Key == "version" {
-								if !*dryRun {
-									err = deregisterImage(ec2Client, image)
-									if err != nil {
-										fmt.Printf("  - error deregistering image %v (%v): %v\n", *image.ImageId, *t.Value, err)
-									} else {
-										fmt.Printf("  - successfully deregistered image %v (%v)\n", *image.ImageId, *t.Value)
+
+					deregisterWg := sync.WaitGroup{}
+					for index, image := range imagesToDelete {
+						wg.Add(1)
+						go func(img ec2Types.Image, imageIndex int) {
+							defer wg.Done()
+							for _, t := range img.Tags {
+								if *t.Key == "version" {
+									var deregisterImageErr error
+									if !(*dryRun) {
+										deregisterImageErr = deregisterImage(ec2Client, img)
 									}
-								} else {
-									fmt.Printf("  - %v (%v)\n", *image.ImageId, *t.Value)
+									drr.deregisterImageResults[imageIndex] = deregisterImageResult{imageId: *img.ImageId, version: *t.Value, err: deregisterImageErr}
+									break
 								}
-								break
 							}
-						}
+						}(image, index)
 					}
+					deregisterWg.Wait()
+
+					deregisterRegionResultChan <- drr
 				} else {
-					fmt.Printf("ERROR: MANUAL ACTION REQUIRED - Unable to delete images in region %v (Total Images: %v, rhel-arm64: %v, legacy-x86_64: %v, rhel-x86_64: %v, Quota: %v, ImagesToDelete: %v)\n",
+					fmt.Printf("ERROR: MANUAL ACTION REQUIRED - Unable to delete images in region %v (Total Images: %v, rhel-arm64: %v, legacy-x86_64: %v, rhel-x86_64: %v, Quota: %v, NumImagesToDelete: %v)\n",
 						regionName, imageCount, len(arm64Images), len(legacyx86Images), len(x86Images), quota, numOfImagesToDelete)
 				}
 			}
 		}(*enabledRegion.RegionName)
 	}
 
 	wg.Wait()
+
+	var deregisterResults []deregisterRegionResult
+	resultsChanLength := len(deregisterRegionResultChan)
```

Comment:
I'd stick with `len(enabledRegions)` here, in case you or some future person decides to adjust `deregisterRegionResultChan` to make it a different length from the number of results we expect it to hold.  At the moment you make it that size, but someone in the future might adjust the amount or presence of buffering in the channel, and not notice that they'd also need to adjust this expected-result-count value down here.

Reply:
My thought is that at this point in the code, we want to process everything in "chan". We'll already have a difference between `len(enabledRegions)` and `len(deregisterRegionResultChan)` because we don't add results if a region is currently under quota.

Curious to hear your thoughts on this though - I'm always interested in optimizing code, especially when it'll be part of our CI process eventually.